### PR TITLE
Move static Index into Mesh

### LINKED ISSUE #1192: Remove static query::Indexer
**Describe your setup**

preCICE 2.3.1

**Describe the problem**

The indexer of the query package internally caches Index trees to prevent them from being regenerated.
This is yet another static state, which we need to remove in order to robustly allow multiple SolverInterfaces.

**Proposed solutions**

One solution would be to attach the index directly to the Mesh.
This would simplify the access to such indexes, as mappings etc already have access to the Mesh.

**Additional context**

#1049

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

This PR removes the `static` caching of the RTrees in the query package and adds the index directly to the `mesh::Mesh`.
This allowed me to remove the use of signals to reset caches etc.
All mappings and watchpoints now use the `mesh::Mesh::index()` instead of using some `static`ally cached entry.

## Motivation and additional information

Less static state and simpler lifetime management.
Using the geometric index is more intuitive, as it can directly be accessed via the `Mesh`.

Closes #1192

## Author's checklist

* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I ran `make format` to ensure everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.10.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: src/mapping/NearestNeighborBaseMapping.cpp [hunk 1]
```diff
@@ -53,11 +51,6 @@ void NearestNeighborBaseMapping::computeMapping()
     searchSpace = input();
   }
 
-  // For log outputs
-  precice::utils::Event e2(baseEvent + ".getIndexOnVertices", precice::syncMode);
-  query::Index          indexTree(searchSpace);
-  e2.stop();
```
Comment (boris-martin) [inline src/mapping/NearestNeighborBaseMapping.cpp:59]: Why remove the event-based stuff?

File: src/mesh/Mesh.cpp [hunk 2]
```diff
@@ -29,18 +29,11 @@ Mesh::Mesh(
...
       _id(id),
-      _boundingBox(dimensions)
+      _boundingBox(dimensions),
+      _index(*this)
 {
   PRECICE_ASSERT((_dimensions == 2) || (_dimensions == 3), _dimensions);
   PRECICE_ASSERT(_name != std::string(""));
-
-  meshChanged.connect([](Mesh &m) { query::clearCache(m); });
-  meshDestroyed.connect([](Mesh &m) { query::clearCache(m); });
```
Comment (boris-martin) [inline src/mesh/Mesh.cpp:38]: When does the clearing happen then? when we call index.clear()?

File: src/precice/impl/WatchPoint.cpp [hunk 3]
```diff
@@ -97,7 +96,7 @@ void WatchPoint::exportPointData(
     return;
   }
 
-  //PRECICE_ASSERT(_vertices.size() == _weights.size());
+  // PRECICE_ASSERT(_vertices.size() == _weights.size());
```
Comment (boris-martin) [inline src/precice/impl/WatchPoint.cpp:99]: Do we want these comments to stay?

File: src/query/Index.hpp [hunk 4]
```diff
@@ -74,22 +75,27 @@ class Index {
...
+  void clear();
+
 private:
-  struct IndexImpl;
+  class IndexImpl;
   std::unique_ptr<IndexImpl> _pimpl;
 
-  const mesh::PtrMesh             _mesh;
+  /// The indexed Mesh.
+  mesh::Mesh *_mesh;
```
Comment (boris-martin) [inline src/query/Index.hpp:97]: Why use raw pointer instead of smart pointers?

**Review by boris-martin — COMMENTED** [review]
Much easier to read overall. I didn't check the details of `Index.cpp` as it looks quite complex, but the impacts on the rest of the code are really nice.

File: src/mapping/NearestNeighborBaseMapping.cpp (hunk 1 above)
Comment (fsimonis) [reply src/mapping/NearestNeighborBaseMapping.cpp:59]: This event is now part of the `Index` itself:
https://github.com/precice/precice/pull/1242/files#diff-4c0c54bc6aa3d8fe12a773602b1c27237290e890cd134b796ca933f85785b012R50

File: src/mesh/Mesh.cpp (hunk 2 above)
Comment (fsimonis) [reply src/mesh/Mesh.cpp:38]: Exactly, this is done in `Mesh::clear()` and when necessary in mappings.

**Comment by fsimonis** [comment]
@boris-martin Thanks for the review!

File: src/query/Index.hpp (hunk 4 above)
Comment (boris-martin) [reply src/query/Index.hpp:97]: Isn't that the purpose of weak pointers?

Comment (fsimonis) [reply src/query/Index.hpp:97]: Weak pointers model a non-owning handle to a shared pointer.
To access the object, you tell the weak pointer that you require ownership, then it returns an owning shared pointer. Think of deferred ownership.

**Pull request merged**